students/timm67/session05/mailroom_part3.py

```diff
@@ -0,0 +1,139 @@
+
+donors = {
+     ('Jim', 'Tillson'): [5.00, 20.00],
+     ('Barb', 'Langley'): [10.00, 20.00, 100.00],
+     ('Jen', 'Garfield'): [10.00, 20.00, 5.00],
+     ('Rex', 'Miller'): [20.00, 20.00, 5.00],
+     ('Tony', 'Blake'): [20.00, 20.00, 10.00]
+     }
+
+exitMail = False
+
+
+def send_thankyou_single_donor_task():
+    global donors
+    donor_found = False
+    donor_fname = input("Enter donor first name ('list' for all donors) : ")
+    # Optimized with List Comprehension
+    if donor_fname == 'list':
+        [print("{}, {} ".format(name[0], name[1])) for name in donors.keys()]
+        donor_fname = input("Enter donor first name : ")
+    donor_lname = input("Enter donor last name : ")
+    while True:
+        donor_amt = input("Enter donation value: ")
+        try:
+            donor_amt = float(donor_amt)
+            break
+        except ValueError:
+            print("Please enter a numeric donation value")
+
+    # Not optimal for optimization via list comprehension
+    for donor in donors.keys():
+        if donor[0] == donor_fname and donor[1] == donor_lname:
+            print("{0}, {1} FOUND".format(donor_fname, donor_lname))
+            donors[donor].append(donor_amt)
+            donor_found = True
+            break
+
+    if not donor_found:
+        print("{0}, {1} NOT found, creating".format(donor_fname, donor_lname))
+        donor = (donor_fname, donor_lname)
+        donors[donor] = [donor_amt]
+
+    fmtline0 = "\n\nDear {0},\n\n"
+    fmtline1 = "Many thanks for your recent donation of {0:6.2f}.\n\n"
+    fmtline2 = "Your total donations are {0:6.2f}, averaging {1:6.2f}\n\n"
+    fmtline3 = "Thanks for your generous donation.\n\nBest Regards,\nThe Staff"
+    donations = donors[donor]
+    total_donations = sum(donations)
+    avg_donation = total_donations / len(donations)
+    print(fmtline0.format(donor[0]))
+    print(fmtline1.format(donations[(len(donations) - 1)]))
+    print(fmtline2.format(total_donations, avg_donation))
+    print(fmtline3.format())
+
+
+def send_thankyou_multiple_donors_task():
+    global donors
+    ffmtline0 = "\n\nDear {0},\n\n"
+    ffmtline1 = "Many thanks for your recent donation of {0:6.2f}.\n\n"
+    ffmtline2 = "Your total donations are {0:6.2f}, averaging {1:6.2f}\n\n"
+    ffmtline3 = "Thanks for your generous donation.\nBest Regards,\nThe Staff"
+    for donor in donors.keys():
+        filename = "./{0}_{1}.txt".format(donor[0], donor[1])
+        print("Writing {0}...".format(filename))
+        donations = donors[donor]
+        total_donations = sum(donations)
+        avg_donation = total_donations / len(donations)
+        with open(filename, 'w') as handle:
+            handle.write(ffmtline0.format(donor[0]))
+            handle.write(ffmtline1.format(donations[(len(donations) - 1)]))
+            handle.write(ffmtline2.format(total_donations, avg_donation))
+            handle.write(ffmtline3.format())
```

Comment:
A good place for `try`-`catch` in case the file writing fails due to errors like `PermissionError` or `IOError`.